HookedProxy.py

```diff
@@ -0,0 +1,413 @@
+        return result
+
+
+class HookedProxyClient(ProxyClient):
+    _response_version = b''
+    _response_code = b''
+    _response_reason = b''
+    _response_headers = {}
+    _response_data_buffer = b''
+    hookChain = HookChain()
+
+    @classmethod
+    def registerHooks(cls, hooks):
+        cls.hookChain.registerHooks(hooks)
+
+    def handleResponsePart(self, data):
+        # Buffer data until all has been read
+        self._response_data_buffer += data
+
+    def handleResponseEnd(self):
+        """
+        Write the buffered headers and body to the transport
+        Finish the original request, indicating that the response has been
+        completely written to it, and disconnect the outgoing transport.
+        """
+        if self._finished:
+            return
+        
+        self._status = f"{self.father.clientproto.decode()} {self.father.code} {self.father.code_message.decode()}".encode('utf-8')
+
+        # Get all headers and save them temporarily
+        for _header, _value in self.father.responseHeaders.getAllRawHeaders():
+            assert len(_value) == 1
+            # For some reason the Headers class stores the header value in a list - extract it
+            self._response_headers[_header] = _value[0]
+        
+        # Also clear the headers from the father so we can write only the returned ones from the hook back
+        # TODO: There is probably a better way of doing this so we don't need two loops, but it's bitching that
+        # I'm changing the size of the iterator mid iteration. rip
+        for _header, _value in self._response_headers.items():
+            self.father.responseHeaders.removeHeader(_header)
+
+        # Apply all hooks
+        self.hookChain.onResponseReceived(
+            self._status, 
+            self._response_headers, 
+            self._response_data_buffer
+        )
+
+        # Father is the request that created the factory that generated this protocol
+        # After applying hook we need to write the headers back to the father's raw headers
+        for k,v in self._response_headers.items():
+            self.father.responseHeaders.addRawHeader(k, v)
+
+        # This will write the headers and the data buffer back to the original client.
+        self.father.write(self._response_data_buffer)
+
+        # Mark it as complete and close the transport
+        self._finished = True
+        self.father.finish()
+        self.transport.loseConnection()
+    
+
+class HookedProxyClientFactory(ProxyClientFactory):
+    """
+    """
+
+    protocol = HookedProxyClient
+    hookChain = HookChain()
+
+    def registerHooks(self, hooks):
+        self.hookChain.registerHooks(hooks)
+        self.protocol.registerHooks(hooks)
+
+
+class HookedReverseProxyRequest(ReverseProxyRequest):
+    ''' This is a request object (encapsulates a request and response)
+    '''
+    proxyClientFactoryClass = HookedProxyClientFactory
+    hookChain = HookChain()
+
+    @classmethod
```

Comment:
Not crazy about these as class methods. Could create weird bugs if you ever tried to create multiple Proxies at once. Can't think of a reason you would do that, but should probably get fixed at some point nonetheless